### PR TITLE
[stable9] Block corrupted filecache updates

### PR DESCRIPTION
**Pull request by tomneedham**
## Description
At some point, the filecache can become corrupted when entries having `fileid=parent`. This PR throws and exception at this point to stop the update and logs a stacktrace to get some logging information as to when this occurs.

## How Has This Been Tested?
 - Creating a folder structure in UI
 - Sharing some files with another user
 - Manually tweaking the parent column of several files to match their id
 - Usage of the UI

Haven't managed to trigger the exception through normal usage.

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

**Comment by SergioBertolinSG** [comment]
The steps are not very clear, are these steps correct?

Is sharing essential here? Where to put a step like "Sharing some files with another user" ?

1. Create folders /a/b
2. Upload 1000.jpg a /a/b/1000.jpg
3. Upload 1473.jpg a /a/b/1473.jpg
4. Get the file ids and parents of uploaded files:

```sql
select fileid from filecache where path='files/a/b';
12

select parent,fileid from filecache where path='files/a/b/1473.jpg';
12 17

select parent,fileid from filecache where path='files/a/b/1000.jpg';
12 15
```
5. Set the parent of these files to match their ids:
```sql
update filecache set parent=15 where path='files/a/b/1000.jpg';
update filecache set parent=17 where path='files/a/b/1473.jpg';
```

6. Using the UI go to /a/b


Following these steps I don't see anything in the logs when browsing the UI.

File: lib/private/files/cache/cache.php [hunk 1]
```diff
@@ -284,6 +285,15 @@ public function insert($file, array $data) {
 	 */
 	public function update($id, array $data) {
 
+		if(isset($data['parent']) && $data['parent'] === $id) {
+			// Catch the case when we are trying to update the parent to be itself
+			\OC::$server->getLogger()->error("Trying to set parent of file: $id to itself!");
+			\OC::$server->getLogger()->error(
```
Comment (PVince81) [inline lib/private/files/cache/cache.php]: put both into the same log entry ?

high concurrency envs could see concurrent entries be inserted between boths which will reduce readability. (can still use request id to track, but why make it so complicated)

**Review by PVince81 — CHANGES_REQUESTED** [review]
looks good, but see comment

**Comment by tomneedham** [comment]
> Following these steps I don't see anything in the logs when browsing the UI.

The trouble here @SergioBertolinSG is that we don't exactly know what is causing the issue, so I cannot trigger the log yet - we just want to catch it if/when it happens

**Review by PVince81 — APPROVED** [review]
:+1:

**Comment by PVince81** [comment]
don't merge yet

**Comment by PVince81** [comment]
Added logging in case of mime type change.

To prevent false positive due to a bug with trashbin, I added some checks to make sure we're not updating mime type when coming from or going to trashbin.

**Comment by PVince81** [comment]
@tomneedham I have just verified that uploading with chunks and part file do not trigger the mime type log warning.

Note that if you purposefully change the extension of a file an entry **will** be logged, but that's not what we're after.

**Comment by tomneedham** [comment]
Closing as we have implemented more appropriate fixes here, including fix for cross storage move info

**Pull request closed without merging**